src/markdown_generators.js

```diff
@@ -159,34 +160,57 @@ const generateMarkdownFromJson = courseData => {
   /**
     This function takes JSON data parsed from a master.json file and returns markdown data
     */
+  this["courseData"] = courseData
+  this["menuIndex"] = 0
```

Comment:
just a thought I had here, in JavaScript 'basic' types like numbers, booleans, strings are pass-by-value when they are passed to functions as arguments, but objects, by contrast are pass-by-reference. so if you do something like this:

```js
let foo = 0
(x => { x += 1 })(foo)
```

then `foo` will still be `0`, because when the function is called it is not passed a reference to the `foo` variable, but instead a new variable `x` is instantiated with a copy of `foo`'s value, so the mutation of `x` doesn't affect the value of `foo`.

but since objects are pass-by-reference, if you do this:

```js
let fooContainer = { foo: 0 }
(x => { x.foo += 1 })(fooContainer)
``` 

then `fooContainer.foo` will be 1, b/c the function receives a reference to the original `fooContainer` variable and not just a copy of it's value, so modifying `fooContainer.foo` in the function body 'escapes' the function's context and modifies the original object.

Anyhow, I think if you wanted to you could use this to refactor away from using `this` to hold the `menuIndex` variable, if it were passed down as `{ menuIndex: 0 }` for instance I think you could modify it anywhere in the recursive function call tree and have it be visible globally elsewhere. Feel free to ignore if you don't want ot bother, just something I was thinking about as I read :smile:

Reply:
I knew about by-ref vs by-val arguments, but didn't didn't know some of these specifics about Javascript.  I tried setting the functions up to take courseData and menuIndex as arguments, but I think it made the code a little messier actually.  Since the default arguments coming through map are the iterated object, its index and the source array... you have to use this:

```
rootSections.map(page => {
      generateMarkdownRecursive({
        page:       page,
        courseData: courseData,
        menuIndex:  0
      })
    })

...

children.map(child => {
      generateMarkdownRecursive({
        page: child,
        courseData: data["courseData"],
        menuIndex: data["menuIndex"]
      })
    })
```

versus just setting the initial data on `this` and not even having to worry about arguments:
```
this["courseData"] = courseData
this["menuIndex"] = 0

rootSections.map(generateMarkdownRecursive, this)

...

children.map(generateMarkdownRecursive, this)
```